src/client/array/dc_array.c

```diff
@@ -2145,17 +2147,20 @@ check_record_cb(tse_task_t *task, void *data)
 
 	/** set memory location */
 	D_ALLOC(val, params->cell_size);
+	if (val == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
 	sgl->sg_nr = 1;
-	D_ALLOC_PTR(sgl->sg_iovs);
+	sgl->sg_iovs = &params->sg_iov;
 	d_iov_set(&sgl->sg_iovs[0], val, params->cell_size);
+	params->user_sgl_used = true;
```

Comment:
just confirm, not very sure for the user_sgl_used flag usage, here it actually uses the params' internal inline sgl/sg_iov, it is not user API task's passed in sgl, right?

Reply:
it is a little confusing. in this case, the calling function here is the user and we are using a preallocated sgl from the struct. it is not needed to allocate 1 here. we set this flag to true so to not free the iov it in the free function cb and we just free the whole args struct.
before we were just allocating and freeing an iov of size 1 which is not ideal (required for async operations to be on the heap not the stack).